pymatgen/io/input.py

```diff
@@ -0,0 +1,172 @@
+"""
+This module defines the abstract interface for pymatgen InputSet and InputSetGenerator classes.
+"""
+
+import abc
+import os
+from pathlib import Path
+from string import Template
+from typing import Union, Optional, Dict
+from zipfile import ZipFile
+from monty.json import MSONable
+from monty.io import zopen
+
+
+__author__ = "Ryan Kingsbury"
+__email__ = "RKingsbury@lbl.gov"
+__status__ = "Development"
+__date__ = "August 2021"
+
+
+class InputSet(MSONable):
```

Comment:
According to the [docs](https://docs.python.org/3/library/collections.abc.html), inheriting from `Set` would mean we have to define `__contains__`, `__len__`, and `__iter__`. It's not clear to me how those concepts map onto what we're discussing here. What are the "items" in this set that will be relevant to all codes? Can you elaborate?

Reply:
E.g., a set of VASP inputs should contain exactly one INCAR, POSCAR, POTCAR and KPOINTS file / corresponding objects. I should be able to iterate the InputSet by doing

```
for i in input_set:
   print(i)

print(len(input_set))   # tells me how many inputs are associated in the set.
```

It is true that `__contains__` is less relevant but even then, it is foreseeable we might want to do:

```
if chgcar in input_set:
   input_set.incar["ICHARG"] = 1
```

Reply:
I'm also confused by this. As I see it, `InputSet` is just an abstraction for one or more inputs with tools for writing and reading them to/from a directory. Subclassing `Set` seems to add an extra layer of complication and I'm not sure what the benefits are? For example, it seems easier to add constraints on the inputs by having an appropriate __init__ method. 

The three complications I see are:
1. Making a class to represent the inputs for a code that only requires one input file will require two classes (in addition to the generators): the "input set" container and the container for the specific input file.
2. All input objects e.g., "Incar" will need to be hashable.
3. Sets are not monty serialisable so this class will require extra care.

I think it will be less hassle if we don't subclass `collections.abc.Set`.

Reply:
Defafult python sets are not serializable, but subclassing abc.Set only requires that you implement the three methods (which are mostly useful in any case). Are there any arguments whether having `__contains__` and `__len__`? If not, then we should subclass abc.Set. Whether they are monty serializable is an implementation of the MSONable protocol. You can always make it monty serializable as long as the underlying objects are serializable.

And finally, all objects are by default hashable unless you make them otherwise e.g., by setting `__hash__ = None` or by inheriting an unhashable object. So it is not too difficult make an input hashable simply by overriding `__hash__`.  In fact, I would argue Incar probably shouldn't inherit dict but rather inherit abc.Mapping instead.

Reply:
An InputSet can be a single file too. Just like Sets can contain a single item. I am just saying that an Input (Incar for example) is one class. An InputSet is another. Just like an integer is a class and a list of integers is another class.

Reply:
> Subclassing Set seems to add an extra layer of complication and I'm not sure what the benefits are?

I agree with @utf here. While I acknowledge that in the case of VASP, our `InputSet` is always clearly a set of underlying objects, that won't be the case for every code. And, more specifically, for codes that only require a single input, I don't want to require developers to implement an extra (e.g. the equivalent of `Incar`) in order to populate `InputSet`.

I'm fully on board with your previous comment @shyuep :

> A final note that I always favor a simple implementation over a complex one. Most other codes are far too serpentine and complex and difficult to use. Do not overengineer something because a 0.01% of people want to do this specialized thing, even if that 0.01% happens to be someone in the Materials Project. If the 0.01% wnat to do something special, they can write their own specialized method. Pymatgen caters to the 80% use case with 20% effort.

So in the interest of keeping this implementation possible, I'd advocate not inheriting from `Set`

Reply:
I don't think you guys are hearing what I am saying. I am not asking to inherit the actual python Set, but rather the abstract base class set. Those are different things. An input set is a set of items. That is the basic definition.

Reply:
Yes, I understand; apologies for being imprecise in my previous post. What I'm saying is that although an `InputSet`  _may_  sometimes be a set of items in a concrete sense (i.e. for VASP, where it contains subsidiary `Incar`, `Potcar`, etc. objects that we define), at other times an `InputSet` may only be a set in the conceptual sense (i.e., there are no subsidiary objects, just a collection of attributes). `TemplateInputSet` is an example of this.

I don't want to require that `InputSet` be populated by other bespoke objects. But in the case where there are no subsidiary objects and the `InputSet` contains all the necessary data as attributes, then I struggle to see how `__contains__` or `__len__` or `__iter__` make sense conceptually.

Reply:
Ok, I don't want to go round and round on this. I have already said that even an InputSet of one item is a set. I also indicated that `__contains__` and `__len__` have actual value. I even gave specific examples. So just implement it. Alternatively, I can merge and I will add it to the code anyway.

Reply:
> Ok, I don't want to go round and round on this. I have already said that even an InputSet of one item is a set. I also indicated that `__contains__` and `__len__` have actual value. I even gave specific examples. So just implement it. Alternatively, I can merge and I will add it to the code anyway.

Yes I appreciate the examples @shyuep and I see what you're getting at; I'm not trying to be dismissive. But those examples only apply when there is a subsidiary object like an `Incar`. At the abstract interface level, I think it should be possible to define an `InputSet` without requiring subsidiary objects (instead, you just populate class attributes as needed). In that case, `InputSet` is not a set of one item; it's a set of zero items. What would you propose in that case?

Reply:
I don't think there is a situation when we have zero items, but even an empty set is still a set. In any case, your example of `TemplateInputSet` is more properly a *generator* under the current definition than an InputSet.

Let's make the concepts clear once and for all. Anything that generate an input, whether it is by rules, or templates, or something else, is an InputGenerator. That is basically the replacement of the *existing* VaspInputSets (and other InputSets). The *new* InputSet contains a set of inputs that have been generated. That is a souped up version of the current VaspInput. InputSets should not be *generating* input on the fly. They are containers.

Reply:
So here is my understanding of the current situation and proposal. Note that I'm sticking with the existing names of `InputSet` and `InputSetGenerator` for now to minimize confusion; open to renaming either though.

### Current Situation

1. Classes that correspond to **single input files** 

- `Incar`, `Poscar`, etc.
- `QCInput`

2. Classes that are **containers** for one or more input files

- `VaspInput`
- `VaspInputSet` / `DictSet(VaspInputSet)`
- `QChemDictSet(QCinput)`

3. Classes that provide **sets of settings for specific calculation types**

- `MPRelaxSet(DictSet)`
- `OptSet(QChemDictSet)`
- etc.

### Proposed Situation

1. Classes that correspond to **single input files** (OPTIONAL)

- `Incar`, `Poscar`, etc.
- `QCInput`

2. Classes that are **containers** for one or more input files

- `InputSet` and child classes
- `VaspInputSet(InputSet)`
- `QChemDictSet(InputSet)`
- `LammpsInputset(InputSet)`

3. Classes that provide **sets of settings for specific calculation types**

- `InputSetGenerator` and child classes
- `MPRelaxSet(InputSetGenerator) -> VaspInputSet`
- `OptSet(InputSetGenerator) -> QChemDictSet`
- `ElectrolyteSet(InputSetGenerator) -> LammpsInputset` 
- etc.

Given the above paradigm, to address @shyuep 's point, I'd propose that we define `InputSet.__len__` as `len(InputSet._generate_input_data.keys()`. Basically, the length is equal to the number of input files that will be written, regardless of how many underlying objects may be attached to the `InputSet`.  `__contains__` would also refer to keys in `InputSet._generate_input_data` (so basically you'd be searching filenames that will be written), e.g. `"INCAR" in VaspInputSet` would evaluate to `True`.

Obviously inheriting classes can overload this behavior.

Reply:
Actually I have some additional thoughts on my previous comment to make `__contains__` more useful; will need some time to implement. Stay tuned!

Reply:
See new comment in the main thread below.